tests/end_to_end_test/snapshot_fail_test.go

```diff
@@ -0,0 +1,140 @@
+package endtoend_test
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"testing"
+
+	"github.com/kopia/kopia/tests/testenv"
+)
+
+func TestSnapshotFail(t *testing.T) {
+	t.Parallel()
+
+	e := testenv.NewCLITest(t)
+	defer e.Cleanup(t)
+	defer e.RunAndExpectSuccess(t, "repo", "disconnect")
+
+	e.RunAndExpectSuccess(t, "repo", "create", "filesystem", "--path", e.RepoDir)
+	e.RunAndExpectSuccess(t, "policy", "set", "--global", "--keep-latest", strconv.Itoa(1<<31-1))
+
+	parentDirOfSource := makeScratchDir(t)
+	source := filepath.Join(parentDirOfSource, "source")
+
+	// Test snapshot of nonexistent directory fails
+	e.RunAndExpectFailure(t, "snapshot", "create", source)
+
+	testenv.MustCreateDirectoryTree(t, source, testenv.DirectoryTreeOptions{
+		Depth:                  2,
+		MaxSubdirsPerDirectory: 1,
+		MaxFilesPerDirectory:   1,
+	})
+
+	// Create snapshot
+	e.RunAndExpectSuccess(t, "snapshot", "create", source)
+
+	numSuccessfulSnapshots := 1
+
+	// Test the root dir permissions
+	for _, tc := range []struct {
+		desc      string
+		parentDir string
+	}{
+		{
+			desc:      "Modify permissions of the snapshot root directory",
```

Comment:
It'd be good to make the test cases as explicit as possible here by hoisting at least the combinations of the outer loop inside `testPermissions`. That is, specify the parameter  if possible instead of computing it.

Currently, for each of these cases there are 2 combinations: dir vs file. (plus the combinations for permissions, but let's ignore those for now).
 - Out out those, there is one case that does not completely make sense, which is that of a file at the same level of the source dir.  Although, that case does not get executed because there is no file when `parentDir == filepath.Dir(source)`

An option is to structure the test (and helper) as
- this is the snapshot source (which is always the same)
- this is the file or directory path for which the permissions would be changed.
- whether a success or a failure is expected.

Then the cases would be:
```go
		{
			desc:      "...",
			modifyPath: filepath.Dir(source),
			success: false,
		},
		{
			desc:      "...",
			modifyPath: source,
			success: false,
		},
		{
			desc:      "...",
			modifyPath: findASubDirIn(source),
			success: false,
		},
		{
			desc:      "...",
			modifyPath: findAFileIn(source),
			success: true,
		},
```

Reply:
Yeah I like this reasoning. Point the test helper at the thing you want to modify permissions on, rather than the directory that contains those things. It's a little more direct